File: TableGames.Games.InBetween/inBetween.js
define(['knockout', 'jquery', 'game!delay', 'tmpl!InBetween/inBetween', 'game!card', 'game!InBetween/inBetweenPlayerHand', 'tmpl!InBetween/inBetweenStats', 'tmpl!dial'], function(ko, $, delay, inBetweenTemplateName, Card, InBetweenPlayerHand, inBetweenStatsTemplateName, dialTemplateName) {

    function InBetween(gameConfig, gameState, table) {
        var self = this;

        self.bank = ko.observableRunner(gameState.bank);
        self.pot = ko.observableRunner(gameState.pot);

        self.playerHands = gameState.playerHands.map(function(playerHandState) { return new InBetweenPlayerHand(playerHandState); });
        self.playerHands.update = function(playerHandsState) {
            playerHandsState.forEach(function(playerHandState, index) {
                self.playerHands[index].updateCards(playerHandState.cards);
                self.playerHands[index].chipsPurchased(playerHandState.chipsPurchased);
                self.playerHands[index].chipsBalance(playerHandState.chipsBalance);
                self.playerHands[index].bet(playerHandState.bet);
                self.playerHands[index].payment(playerHandState.payment);
            });
        };

        table.activePlayerName(gameState.table.activePlayerName);
    }
    InBetween.templateName = inBetweenTemplateName;
    InBetween.statsTemplateName = inBetweenStatsTemplateName;
    InBetween.LOT_AMOUNT = 500;
    InBetween.MIN_BET_AMOUNT = 10;

    InBetween.ActionConstructors = [
        function BetAction(gameConfig, game, table) {
            var self = this;
            self.name = 'bet';

            var userHand = ko.utils.arrayFirst(game.playerHands, function(ph) { return ph.playerName === game.table.activePlayerName(); });

            // max bet
            self.maxBet = ko.computed(function() { return Math.min(userHand.chipsBalance() / 2, game.pot()); });

            // bet amount
            self.amount = ko.observable(InBetween.MIN_BET_AMOUNT);
            self.amount.dial = {
                up: {
                    execute: function() {
                        if (self.amount() < self.maxBet()) {
                            self.amount(self.amount() + InBetween.MIN_BET_AMOUNT);
                        }
                    },
                    enable: table.isUserTurn
                },
                down: {
                    execute: function() {
                        if (self.amount() > InBetween.MIN_BET_AMOUNT) {
                            self.amount(self.amount() - InBetween.MIN_BET_AMOUNT);
                        }
                    },
                    enable: table.isUserTurn
                }
            };
            self.amount.setMaxBet = function() { self.amount(self.maxBet()); };
            self.amount.setMinBet = function() { self.amount(InBetween.MIN_BET_AMOUNT); };

            self.getParameters = function() {
                return {
                    amount: self.amount()
                };
            };

            self.onExecuted = function(playerName, gameChangeResults) {
                var userHandChangeResults = ko.utils.arrayFirst(gameChangeResults.playerHands, function(ph) { return ph.playerName === game.table.activePlayerName(); });

                var promise = $.when()
                                .then(function() {  // update player bet, remove bet from player chips
                                    var betPromise = userHand.bet.runTo(userHandChangeResults.bet);
                                    var chipsPromise = userHand.chipsBalance.runTo(userHand.chipsBalance() - userHandChangeResults.bet);
                                    return $.when(betPromise, chipsPromise);
                                })
                                .then(function() {
                                    userHand.updateCards(userHandChangeResults.cards);
                                    return delay();
                                })
                                .then(function() { // if win
                                    if (userHandChangeResults.payment === userHandChangeResults.bet) {
                                        // pot => bet
                                        var potPromise = game.pot.runTo(gameChangeResults.pot);
                                        var betPromise = userHand.bet.runTo(2 * userHandChangeResults.bet);
                                        var winPromise = $.when(potPromise, betPromise).then(function() {
                                            // bet => chips
                                            var betPromise = userHand.bet.runTo(0);
                                            var chipsPromise = userHand.chipsBalance.runTo(userHandChangeResults.chipsBalance);
                                            return $.when(betPromise, chipsPromise);
                                        });
                                        return winPromise;
                                    }
                                })
                                .then(function() { // if simple lose
                                    if (userHandChangeResults.payment === -userHandChangeResults.bet) {
                                        // bet => pot
                                        var betPromise = userHand.bet.runTo(0);
                                        var potPromise = game.pot.runTo(gameChangeResults.pot);
                                        return $.when(potPromise, betPromise);
                                    }
                                })
                                .then(function() { // if post lose
                                    if (userHandChangeResults.payment === -2 * userHandChangeResults.bet) {
                                        // chips => bet
                                        var chipsPromise = userHand.chipsBalance.runTo(userHandChangeResults.chipsBalance);
                                        var betPromise = userHand.bet.runTo(2 * userHandChangeResults.bet);
                                        var losePromise = $.when(chipsPromise, betPromise).then(function() {
                                            // bet => pot
                                            var betPromise = userHand.bet.runTo(0);
                                            var potPromise = game.pot.runTo(gameChangeResults.pot);
                                            return $.when(betPromise, potPromise);
                                        });
                                        return losePromise;
                                    }
                                })
                                .then(function() {
                                    userHand.payment(userHandChangeResults.payment);
                                    game.isEnded(gameChangeResults.isEnded);
                                    table.status(gameChangeResults.table.status);
                                    table.activePlayerName(gameChangeResults.table.activePlayerName);
                                    table.stats(gameChangeResults.table.stats);
                                });
                return promise;
            };
        },
        function BuyAction(gameConfig, game, table) {
            var self = this;
            self.name = 'buy';

            var userHand = ko.utils.arrayFirst(game.playerHands, function(ph) { return ph.playerName === game.table.activePlayerName(); });

            self.returnChips = ko.observable(false);
            self.returnChips.enable = ko.computed(function() { return table.isUserTurn() && userHand.chipsBalance() > InBetween.LOT_AMOUNT; });

            self.getParameters = function() {
                return {
                    returnChips: self.returnChips()
                };
            };

            self.onExecuted = function(playerName, gameChangeResults) {
                var userHandChangeResults = ko.utils.arrayFirst(gameChangeResults.playerHands, function(ph) { return ph.playerName === game.table.activePlayerName(); });

                var purchasedPromise = userHand.chipsPurchased.runTo(userHandChangeResults.chipsPurchased);
                var chipsPromise = userHand.chipsBalance.runTo(userHandChangeResults.chipsBalance);
                var bankPromise = game.bank.runTo(gameChangeResults.bank);

                var promise = $.when(purchasedPromise, chipsPromise, bankPromise).then(function() {
                    self.returnChips(false);    // reset returnChips checkbox
                });

                return promise;
            };
        }
    ];

    return InBetween;
});

File: TableGames.Games.InBetween/InBetween.cs
using System;
using System.Collections.Generic;
using System.Linq;
using TableGames.Domain;
using TableGames.Domain.Extensions;

namespace TableGames.Games.InBetween
{
    [GameDescriptor("In Between", InitialGameType = typeof(HigherCard))]
    public class InBetween : Game
    {
        public const int LOT_AMOUNT = 500;
        public const int MIN_BET_AMOUNT = 10;

        private readonly Chips _chips;
        private readonly List<Deck> _decks;

        public List<InBetweenPlayerHand> PlayerHands { get; set; }
        public InBetweenPlayerHand ActiveHand { get; private set; }
        public Deck Deck {
            get { return _decks.Last(); }
        }
        public int Bank {
            get { return _chips.Amount; }
            set { _chips.Amount = value; }
        }
        public int Pot {
            get { return _chips.Balance; }
            set { _chips.Balance = value; }
        }

        public InBetween(Table table) : base(table) {
            _chips = Table.GetBagItem("TablePot", () => new Chips());
            _decks = Table.GetBagItem("TableDecks", () => new List<Deck>() { new Deck() });

            if (Deck.Count < 3) {
                _decks.Add(new Deck());
            }

            var playerBags = Table.GetBagItem("PlayerBags", () => Table.Players.ToDictionary(
                p => p,
                p => new InBetweenPlayerBag()
            ));
            PlayerHands = new List<InBetweenPlayerHand>(playerBags.Select(kvp => new InBetweenPlayerHand(kvp.Key, kvp.Value)));

            // initial pot
            if (Pot == 0) {
                foreach (var playerHand in PlayerHands) {
                    if (playerHand.ChipsBalance <= 3 * MIN_BET_AMOUNT) {
                        playerHand.ChipsPurchased += LOT_AMOUNT;
                        playerHand.ChipsBalance += LOT_AMOUNT;
                        Bank += LOT_AMOUNT;
                    }

                    playerHand.ChipsBalance -= MIN_BET_AMOUNT;
                    Pot += MIN_BET_AMOUNT;
                }
            }

            // first game of series
            if (Table.ActivePlayer == null) {
                Table.SetNextPlayer();
            }
            else {
                if (IsEliminated(Table.ActivePlayer)) {
                    Table.SetNextPlayer();
                }
            }

            // deal active hand
            ActiveHand = PlayerHands.First(ph => ph.Player == Table.ActivePlayer);
            ActiveHand.Cards.Add(Deck.DealExposed());
            ActiveHand.Cards.Add(Deck.DealExposed());
        }

        public override bool IsEliminated(Player player) {
            return PlayerHands.First(ph => ph.Player == player).ChipsBalance <= 0;
        }

        public override object ToClient() {
            var client = new {
                bank = Bank,
                pot = Pot,
                playerHands = PlayerHands.Select(pc => pc.ToClient()),
            };
            return client.Merge(base.ToClient());
        }

        public override object ToStats() {
            var stats = new {
                playerName = ActiveHand.Player.Name,
                bet = ActiveHand.Bet,
                payment = ActiveHand.Payment
            };
            return stats.Merge(base.ToStats());
        }
    }

    public class BetAction : GameAction
    {
        public override string Name => "bet";

        private InBetween _inBetween;

        public BetAction(InBetween inBetween) {
            _inBetween = inBetween;
        }

        public GameChangeResult Execute(int amount) {
            var maxBetAmount = Math.Max(_inBetween.ActiveHand.ChipsBalance / 2, _inBetween.Pot);
            if (amount > maxBetAmount) {
                throw new TableGamesException("Invalid Bet amount.");
            }

            _inBetween.ActiveHand.Bet = amount;
            _inBetween.ActiveHand.Cards.Add(_inBetween.Deck.DealExposed());

            var leftValue = _inBetween.ActiveHand.Cards[0].ValueA1;
            var rightValue = _inBetween.ActiveHand.Cards[1].ValueA1;
            var middleValue = _inBetween.ActiveHand.Cards[2].ValueA1;

            if ((leftValue < middleValue && middleValue < rightValue) || (leftValue > middleValue && middleValue > rightValue)) {
                _inBetween.ActiveHand.Payment = amount;        // win
                _inBetween.ActiveHand.Cards[2].IsHighlighted = true;
            }
            else if (leftValue == middleValue || middleValue == rightValue) {
                _inBetween.ActiveHand.Payment = -2 * amount;   // post lose
            }
            else {
                _inBetween.ActiveHand.Payment = -amount;       // simple lose
            }
            _inBetween.ActiveHand.ChipsBalance += _inBetween.ActiveHand.Payment;
            _inBetween.Pot -= _inBetween.ActiveHand.Payment;

            _inBetween.IsEnded = true;

            _inBetween.Table.SetNextPlayer();

            var gameState = new {
                playerHands = _inBetween.PlayerHands.Select(pc => pc.ToClient()),
                pot = _inBetween.Pot,
                isEnded = _inBetween.IsEnded,
                table = new {
                    status = _inBetween.Table.Status.ToString(),
                    activePlayerName = _inBetween.Table.ActivePlayer?.Name,
                    stats = _inBetween.Table.GetStats(),
                }
            };

            return new GameChangeResult(gameState);
        }
    }

    public class BuyAction : GameAction
    {
        public override string Name => "buy";

        private InBetween _inBetween;

        public BuyAction(InBetween inBetween) {
            _inBetween = inBetween;
        }

        public GameChangeResult Execute(bool returnChips) {
            if (returnChips) {
                if (_inBetween.ActiveHand.ChipsBalance < InBetween.LOT_AMOUNT) {
                    throw new TableGamesException($"Chips Balance is less than {InBetween.LOT_AMOUNT}");
                }
            }

            var factor = returnChips ? -1 : 1;

            _inBetween.ActiveHand.ChipsBalance += factor * InBetween.LOT_AMOUNT;
            _inBetween.ActiveHand.ChipsPurchased += factor * InBetween.LOT_AMOUNT;
            _inBetween.Bank += factor * InBetween.LOT_AMOUNT;

            var gameState = new {
                playerHands = _inBetween.PlayerHands.Select(pc => pc.ToClient()),
                bank = _inBetween.Bank,
                table = new {
                    stats = _inBetween.Table.GetStats(),
                }
            };

            return new GameChangeResult(gameState);
        }
    }

}